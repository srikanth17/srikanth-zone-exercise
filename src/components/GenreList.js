import React from 'react'
import PropTypes from 'prop-types'

export const GenreList = ({ genres, toggleCheckbox }) => (
    <div className="row">
        <h4>Genres</h4>
        {genres.map((genre) => (
            <p key={genre.id} className="col s12 m6 l4">
                <label>
                    <input
                        name="genre"
                        type="checkbox"
                        className="filled-in"
                        onClick={toggleCheckbox}
                        value={genre.id}
                    />
                    <span>{genre.name}</span>
                </label>
            </p>
        ))}
    </div>
);

GenreList.propTypes = {
    genres: PropTypes.array.isRequired,
    toggleCheckbox: PropTypes.func.isRequired
};