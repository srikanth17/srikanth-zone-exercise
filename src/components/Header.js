import React from 'react'

export default () => (
    <div className="header">
        <div className="container">
            <h1 className="header-title">Zone Application</h1>
        </div>
    </div>
);