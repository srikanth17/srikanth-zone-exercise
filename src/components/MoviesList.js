import React from 'react'
import PropTypes from 'prop-types'
import { findGenres } from '../utils/utils'

export const MoviesList = ({ movies, images, genres }) => {
    const baseUrl = images.base_url;
    const imageSize = 'w185'; // Todo, retrieve this from images object
    return (
        <div className="row">
            {movies.map(movie => (
                <div key={movie.id} className="col s12 m6 l4">
                    <div className="card">
                        <div className="card-image">
                            <img src={baseUrl+imageSize+movie.poster_path} />
                            <span className="card-title">{movie.title}</span>
                        </div>
                        <div className="card-action">
                            {findGenres(movie.genre_ids, genres)}
                        </div>
                    </div>
                </div>
            ))}
        </div>
    );
};

MoviesList.propTypes = {
    movies: PropTypes.array.isRequired,
    images: PropTypes.object.isRequired,
    genres: PropTypes.array.isRequired
};