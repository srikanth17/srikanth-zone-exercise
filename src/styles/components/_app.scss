.header {
  background: $dark-blue;
  padding: $header-space;
  margin-bottom: $header-space;
}

.header-title {
  color: white;
  margin: 0;
}

.waves-effect {
  margin-right: 10px;
}