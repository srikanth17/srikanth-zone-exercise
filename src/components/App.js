import React from 'react'
import axios from 'axios'
import { intersection } from 'lodash'
import Header from './Header'
import { GenreList } from './GenreList'
import { MoviesList } from './MoviesList'

class App extends React.Component {
    state = {
        range: 3,
        movies: { data: { results: [] }},
        nowPlaying: { data: { results: [] }},
        genres: { data: { genres: [] }},
        image: { data: { images: {} }},
        selected: []
    };
    componentDidMount() {
        const API_KEY = '00393c8e8abcc530b2d0516bc3304e54';
        axios.all([
            axios.get(`https://api.themoviedb.org/3/movie/now_playing?api_key=${API_KEY}&language=en-UK&page=1`),
            axios.get(`https://api.themoviedb.org/3/genre/movie/list?api_key=${API_KEY}&language=en-UK`),
            axios.get(`https://api.themoviedb.org/3/configuration?api_key=${API_KEY}`)
        ])
            .then(axios.spread((nowPlaying, genres, image) => {
                this.setState({
                    nowPlaying,
                    genres,
                    image
                }, () => {
                    this.setState({
                        movies: JSON.parse(JSON.stringify(this.state.nowPlaying)) // deep copying                        
                    });
                });
            }));
    }

    addRange = () => {
        this.setState({
            range: this.state.range+0.5
        }, this.fetchUpdatedMovies);
    };

    minusRange = () => {
        this.setState({
            range: this.state.range-0.5
        }, this.fetchUpdatedMovies);
    };

    resetRange = () => {
        this.setState({
            range: 3
        }, this.fetchUpdatedMovies);
    };

    // filter and update the movies depending upon user actions
    fetchUpdatedMovies = () => {
        const { nowPlaying, selected, range } = this.state;
        const newMovies = nowPlaying;
        newMovies.data.results = nowPlaying.data.results.filter(movie => {
            return movie.vote_average >= range && intersection(movie.genre_ids, selected).length === selected.length;
        });
        this.setState({
            movies: newMovies
        });
    };

    // handles genre selections
    toggleCheckbox = e => {
        if(this.state.selected.includes(parseInt(e.target.value))) {
            const index = this.state.selected.indexOf(parseInt(e.target.value));
            const newSelected = this.state.selected;
            if(index > -1) {
                newSelected.splice(index, 1);
                this.setState({
                    selected: newSelected
                }, this.fetchUpdatedMovies);
            }
        } else {
            this.setState({
                selected: [...this.state.selected, parseInt(e.target.value)]
            }, this.fetchUpdatedMovies);
        }
    };    

    render () {        
        const {
            range,
            genres,
            movies,
            image,
        } = this.state;
        return (
            <div>
                <Header/>
                <div className="container">
                    <div className="row">
                        <div className="col s12">
                            <h4>Range: {range}</h4>
                            <a className="waves-effect waves-light btn-small" onClick={this.addRange}>Increment</a>
                            <a className="waves-effect waves-light btn-small" onClick={this.minusRange}>Decrement</a>
                            <a className="waves-effect waves-light btn-small" onClick={this.resetRange}>Default</a>
                        </div>
                        <div className="col s12">                            
                            <GenreList
                                genres={genres.data.genres}
                                toggleCheckbox={this.toggleCheckbox}
                            />                                                            
                            <MoviesList
                                movies={movies.data.results}
                                images={image.data.images}
                                genres={genres.data.genres}
                            />
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

export default App;