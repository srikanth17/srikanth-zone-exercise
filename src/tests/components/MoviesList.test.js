import React from 'react'
import { shallow } from 'enzyme'
import { MoviesList } from '../../components/MoviesList'

let movies, images, genres, wrapper;

beforeEach(() => {
    movies = [{
        id: 1,
        title: 'Title1',
        genre_ids: [1, 2],
        poster_path: '/abc'
    }, {
        id: 2,
        title: 'Title2',
        genre_ids: [3, 4],
        poster_path: '/def'
    }];
    images = {
        base_url: 'www.example.com/'
    };
    genres = [{
        id: 1,
        name: "Action"
    }, {
        id: 2,
        name: "Adventure"
    }, {
        id: 3,
        name: "Animation"
    }, {
        id: 4,
        name: "Comedy"
    }, {
        id: 5,
        name: "Crime"
    }];
    wrapper = shallow(<MoviesList movies={movies} images={images} genres={genres}/>);
});

test('should render MoviesList correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should find the correct src prop', () => {
    expect(wrapper.find('img').at(0).prop('src')).toBe('www.example.com/w185/abc');
});