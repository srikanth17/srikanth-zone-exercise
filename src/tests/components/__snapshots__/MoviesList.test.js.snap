// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MoviesList correctly 1`] = `
<div
  className="row"
>
  <div
    className="col s12 m6 l4"
    key="1"
  >
    <div
      className="card"
    >
      <div
        className="card-image"
      >
        <img
          src="www.example.com/w185/abc"
        />
        <span
          className="card-title"
        >
          Title1
        </span>
      </div>
      <div
        className="card-action"
      >
        Action | Adventure
      </div>
    </div>
  </div>
  <div
    className="col s12 m6 l4"
    key="2"
  >
    <div
      className="card"
    >
      <div
        className="card-image"
      >
        <img
          src="www.example.com/w185/def"
        />
        <span
          className="card-title"
        >
          Title2
        </span>
      </div>
      <div
        className="card-action"
      >
        Animation | Comedy
      </div>
    </div>
  </div>
</div>
`;
